contracts/contracts/UserReplicaSetManager.sol

```diff
@@ -0,0 +1,395 @@
+    bytes32 private userFactoryRegistryKey;
+
+    // Address permissioned to update user replica sets
+    address userReplicaSetBootstrapAddress;
+
+    // spID to ServiceProvider delegateOwnerWallet
+    mapping (uint => address) spIdToContentNodeDelegateWallet;
+
+    // Struct used to represent replica sets
+    // Each uint reprsets the spID registered on eth-contracts
+    struct ReplicaSet {
+        uint primary;
+        uint[] secondaries;
+    }
+
+    // Consider frontrunning of spIDs
+    uint[] bootstrapSPIds;
+    // TODO: Do we even need this second array? We just need ONE
+    address[] bootstrapNodeDelegateWallets;
+
+    // Current uint userId to Replica Set
+    mapping (uint => ReplicaSet) userReplicaSets;
+
+    /* Events */
+    event UpdateReplicaSet(
+        uint _userId,
+        uint _primary,
+        uint[] _secondaries
+    );
+
+    event AddOrUpdateContentNode(
+        uint _newCnodeId,
+        address _newCnodeDelegateOwnerWallet,
+        uint[3] _proposerSpIds,
+        address _proposer1Address,
+        address _proposer2Address,
+        address _proposer3Address
+    );
+
+    /* EIP-712 */
+    bytes32 constant PROPOSE_ADD_UPDATE_CNODE_REQUEST_TYPEHASH = keccak256(
+        "ProposeAddOrUpdateContentNode(uint newCnodeId,address newCnodeDelegateOwnerWallet,uint proposerSpId,bytes32 nonce)"
+    );
+    bytes32 constant UPDATE_REPLICA_SET_REQUEST_TYPEHASH = keccak256(
+        "UpdateReplicaSet(uint userId,uint primary,bytes32 secondariesHash,uint oldPrimary,bytes32 oldSecondariesHash,bytes32 nonce)"
+    );
+
+    function initialize(
+        address _registryAddress,
+        bytes32 _userFactoryRegistryKey,
+        address _userReplicaSetBootstrapAddress,
+        uint[] memory _bootstrapSPIds,
+        address[] memory _bootstrapNodeDelegateWallets,
+        uint _networkId
+    ) public initializer
+    {
+        require(
+            _registryAddress != address(0x00) &&
+            _userFactoryRegistryKey.length != 0,
+            "Requires non-zero _registryAddress and registryKey"
+        );
+        registry = RegistryInterface(_registryAddress);
+        userFactoryRegistryKey = _userFactoryRegistryKey;
+        userReplicaSetBootstrapAddress = _userReplicaSetBootstrapAddress;
+        _seedBootstrapNodes(_bootstrapSPIds, _bootstrapNodeDelegateWallets);
+        // Initialize base Signing Logic contract
+        SigningLogicInitializable.initialize(
+            "User Replica Set Manager",
+            "1",
+            _networkId
+        );
+    }
+
+    // Chain of trust based authentication scheme
+    // Nodes are required to have an identity in Audius L2 and this function enables
+    //     known entities to register other known entities on L2 contracts.
+    //  Multiple distinct parties must sign and submit as part of this request
+    function addOrUpdateContentNode(
+        uint _newCnodeId,
+        address _newCnodeDelegateOwnerWallet,
```

Comment:
just thinking through this scenario. if someone updates their delegateOwner wallet, this function can be called to submit their new spId, delegateWallet combo with 3 signers and it will override the old spIdToContentNodeDelegateWallet[spId] with the new delegateWallet

Reply:
another scenario. if someone de-registers and re-registers their existing service so they get a new spId, essentially this proof will have to be resubmitted, but because they lose the spId association in all the structs, it's as if that node just went away and the re-registered one is effectively brand new even though it may have some state on it.

Reply:
also is this the right function to call in either of these scenarios?

Reply:
for the 2nd comment, isn't that state literally transferred to the new spID if you de-register and re-register

Reply:
for the 1st comment, yes that would make perfect sense - assuming the bootstrap nodes don't change their delegateWallet this should work as expected

Reply:
and yes - I do think this would be the right function to call. however these scenarios need tested

Reply:
@hareeshnagaraj the db and file system state is preserved if you re-register but since the spId is different, won't associations in this contract on chain effectively be orphaned?

Reply:
oh yes, we would have to come back and update the spID->delegateWallet tuple on chain

Reply:
One more thing. On what condition does the cnode check its state against this contract and call this function? To me seems like definitely on service unit, but also maybe on an interval like each SnapBack run

Reply:
great point, I agree that it should _at least_ be every Snapback run so we can update the 'signer' response field and maintain client verification